workflow/airflow/dags/local_file_sensor_dag.py

```diff
@@ -0,0 +1,94 @@
+import os
+
+import airflow
+from airflow import DAG
+from airflow.operators.python_operator import PythonOperator
+from airflow.operators.sensors import BaseSensorOperator
+from datetime import timedelta
+
+
+class FileWatcherOperator(BaseSensorOperator):
+    """
+    Triggers a file system check. Parses a file if present.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize a FileWatcherOperator which should wait for a file to be present in a target directory.
+
+        :param args:
+        :param kwargs:
+        """
+        self.directory_path = kwargs.get('directory_path')
+        super(FileWatcherOperator, self).__init__(*args, **kwargs)
+
+    def poke(self, context, *args, **kwargs):
+        """
+        See if a file is present in a given directory.
+
+        :param context: Context dictionary forwarded from the DAG;
+        expects to contain context.dag and context.task
+        :type context: dict
+        :param args: not expected
+        :param kwargs: not expected
+        :return: whether the poke method has succeeded or not, 0 or 1
+        :rtype: int
+        """
+        path = os.path.join(os.getcwd(), self.directory_path)
+        dir_listing = os.listdir(path)
+
+        if len(dir_listing) > 0:
+            context['ti'].xcom_push(key='file_path', value=os.path.join(path, dir_listing[0]))
+            return True
+
+        return False
+
+
+DIR_PATH = 'data'
+
+
+def parse_file(*args, **kwargs):
+    """Simply parse the first file in a directory and return it's contents.
+
+    :param args:
+    :param kwargs:
+    :return:
+    """
+    # get file path value from passed in context
+    file_path = kwargs['ti'].xcom_pull(task_ids='file_watcher', key='file_path')
+
+    with open(file_path) as f:
+        return f.read()
```

Comment:
so once a value is returned, where does it go? In general we tend to have "void" activities that have some side-effect on an external resource (push some files somewhere, call an API, put a message into a queue) so what use cases are return values good for?

Reply:
In this case its available for other operators to access via XCOM, but we could just as well have executed a side effect within the call and not bothered with returning the value. They can just be "void" activities/tasks if you want them to be.